### PR TITLE
ENH: Expression template

### PR DESCRIPTION
**Pull request by erdieee**
## Summary

Add matrix and vector expressions

## Quick changelog

Vector and Matrix classes have be converterd into expressions for fast and elegance computation.
Matrix inverse is not working correctly this point of time. Should also probably be implemented better.

## TODO:

- [x] Matrix Matrix addition
- [x] Matrix Matrix multiplication
- [x] Matrix Vector multiplication
- [x] Vector Vector multiplication
- [x] view row, col, rows, cols of matrix
- [x] Matrix inverse

### CODE REVIEW

File: src/expression.h [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+    template <typename TM, typename TV>
+    class MatVecExpr : public VecExpr<MatVecExpr<TM, TV>>
+    {
+        TM m_;
+        TV v_;
+
+    public:
+        MatVecExpr(TM m, TV v) : m_(m), v_(v) {}
+        auto Upcast() { return MatVecExpr(m_, v_); }
```
Comment (shirnschall) [inline src/expression.h:133]: is Upcast() needed and correct here?

File: src/expression.h [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    template <typename TSCAL, typename TM>
+    class ScaleMatExpr : public MatExpr<ScaleMatExpr<TSCAL, TM>>
+    {
+        TSCAL scal_;
+        TM mat_;
+
+    public:
+        ScaleMatExpr(TSCAL scal, TM mat) : scal_(scal), mat_(mat) {}
+        auto operator()(size_t i, size_t j) const { return scal_ * mat_(i, j); }
+        size_t nRows() const { return mat_.nRows(); }
```
Comment (shirnschall) [inline src/expression.h:115]: functions should be capitalized

File: src/expression.h [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+    template <typename TA, typename TB>
+    class MatMatExpr : public MatExpr<MatMatExpr<TA, TB>>
+    {
+        TA m1_;
+        TB m2_;
+
+    public:
+        MatMatExpr(TA m1, TB m2) : m1_(m1), m2_(m2) {}
+        auto Upcast() { return MatVecExpr(m1_, m2_); }
```
Comment (shirnschall) [inline src/expression.h:155]: returns matvecexpression and is probably not needed

File: src/matrix.h [hunk 4]
```diff
@@ -1,10 +1,10 @@
-#ifndef FILE_Matrix_H
-#define FILE_Matrix_H
+#ifndef FILE_MATIRX_H
```
Comment (shirnschall) [inline src/matrix.h:1]: "matirx" typo

File: src/matrix.h [hunk 5]
```diff
@@ -15,107 +15,103 @@ namespace bla
         ColMajor,
         RowMajor
     };
-    template <typename T, ORDERING ORD = ORDERING::ColMajor>
-    class Matrix
+    template <typename T, ORDERING ORD>
```
Comment (shirnschall) [inline src/matrix.h:18]: no default ordering type

File: src/matrix.h [hunk 6]
```diff
@@ -15,107 +15,103 @@ namespace bla
...
+    class MatrixView : public MatExpr<MatrixView<T, ORD>>
     {
-        size_t rows_, cols_;
-        std::shared_ptr<T[]> data_;
-        bool isTransposed_ = false;
+    protected:
+        size_t rows_, cols_, dist_;
```
Comment (shirnschall) [inline src/matrix.h:22]: dist_ should be a integral constant 1 if unused? (see vector.h)

File: src/matrix.h [hunk 7]
```diff
@@ -15,107 +15,103 @@ namespace bla
...
-        {
-        }
-
-        Matrix(size_t rows, size_t cols, std::shared_ptr<T[]> data, bool isTransposed)
-            : rows_(rows), cols_(cols), data_(data)
-        {
-            isTransposed_ = isTransposed ? false : true;
-        }
-        Matrix(const Matrix &m) : Matrix(m.NumRows(), m.NumCols()) { *this = m; }
+        MatrixView(size_t rows, size_t cols, size_t dist, T *data)
```
Comment (shirnschall) [inline src/matrix.h:26]: dist as template argument see comment above

File: src/matrix.h [hunk 8]
```diff
@@ -15,107 +15,103 @@ namespace bla
...
-            std::swap(data_, m.data_);
-        }
-
-        ~Matrix() {}
-
-        Matrix &operator=(const Matrix &v2)
-        {
-            for (size_t i = 0; i < DataSize(); i++)
-                data_[i] = v2(i);
+            for (size_t i = 0; i < rows_; i++)
```
Comment (shirnschall) [inline src/matrix.h:32]: do not access private/protected variables directly. use nRows()

File: src/matrix.h [hunk 9]
```diff
@@ -15,107 +15,103 @@ namespace bla
...
+        const T *Data() const { return data_; }
         T &operator()(size_t i) { return data_[i]; }
         const T &operator()(size_t i) const { return data_[i]; }
         T &operator()(size_t i, size_t j)
         {
             if (ORD == RowMajor)
-            {
-                return IsTransposed() ? Data()[j * NumCols() + i] : Data()[i * NumCols() + j];
-            }
+                return data_[i * dist_ + j];
```
Comment (shirnschall) [inline src/matrix.h:55]: missing nCols()?

File: src/matrix.h [hunk 10]
```diff
@@ -15,107 +15,103 @@ namespace bla
...
-            {
-                return IsTransposed() ? Data()[j + i * NumRows()] : Data()[i + j * NumRows()];
-            }
+                return data_[i + j * dist_];
+        }
+
+        auto Row(size_t i) const
+        {
+            if constexpr (ORD == RowMajor)
+                return VectorView<T>(cols_, data_ + i * dist_);
```
Comment (shirnschall) [inline src/matrix.h:70]: what is supposed to be stored in Matrix dist_?

File: src/matrix.h [hunk 11]
```diff
@@ -15,107 +15,103 @@ namespace bla
...
+            else
+                return VectorView<T, size_t>(cols_, dist_, data_ + i);
```
Comment (shirnschall) [inline src/matrix.h:72]: replace dist_ with nRows()?

File: src/matrix.h [hunk 12]
```diff
@@ -15,107 +15,103 @@ namespace bla
...
+            if constexpr (ORD == ColMajor)
+                return VectorView<T>(rows_, data_ + i * dist_);
+            else
+                return VectorView<T, size_t>(rows_, dist_, data_ + i);
+        }
+
+        auto Rows(size_t first, size_t next) const
+        {
+            if constexpr (ORD == ColMajor)
+                return MatrixView(next - first, cols_, dist_, data_ + first);
```
Comment (shirnschall) [inline src/matrix.h:85]: please explain this part in more detail

File: src/matrix.h [hunk 13]
```diff
@@ -15,107 +15,103 @@ namespace bla
...
+                return MatrixView(rows_, next - first, dist_, data_ + first * dist_);
         }
 
-        void RowAddRow(size_t row, size_t to)
+        auto Transpose()
         {
-            for (size_t i = 0; i < NumCols(); i++)
-                (*this)(to, i) += (*this)(row, i);
+            if constexpr (ORD == RowMajor)
+                return MatrixView<T, ColMajor>(nCols(), nRows(), nCols(), Data());
```
Comment (shirnschall) [inline src/matrix.h:101]: why pass nCols() twice?

**Review by shirnschall — CHANGES_REQUESTED** [review]
As these are rather large changes, please explain the code in detail in person before merging.

File: src/matrix.h (hunk 6 above)
Comment (erdieee) [reply src/matrix.h:22]: no, dist_ is either the number of rows or cols depending on the ordering.

**Review by shirnschall — APPROVED** [review]

**Comment by shirnschall** [comment]
Works for now 🚀

**Pull request merged**